### PR TITLE
feat: auto download protoc

### PR DESCRIPTION
**Pull request by ShiKaiWi**
Currently, the dependency of protoc has to be downloaded by the users, leading to many troubles for building their project on github.

Let's make the protoc can be downloaded automatically.

### CODE REVIEW

File: rust/build.rs [hunk 1]
```diff
@@ -1,6 +1,10 @@
 // Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
 
 fn main() -> Result<(), Box<dyn std::error::Error>> {
+    // Download the protoc and set the path for tonic_build.
+    let protoc_path = protoc_bin_vendored::protoc_bin_path().unwrap();
+    std::env::set_var("PROTOC", protoc_path.as_os_str());
```
Comment (Rachelint) [inline rust/build.rs:6]: If `protoc` actually exists in local? should we avoid downloading?
Maybe we should check it at first, for example:
```
 if let Err(_) = which::which("protoc") {
      download
}
```

File: rust/build.rs [hunk 2]
```diff
@@ -1,6 +1,10 @@
 // Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
 
 fn main() -> Result<(), Box<dyn std::error::Error>> {
+    // Download the protoc and set the path for tonic_build.
+    let protoc_path = protoc_bin_vendored::protoc_bin_path().unwrap();
```
Comment (Rachelint) [inline rust/build.rs:5]: How about:
```suggestion
    let protoc_path = protoc_bin_vendored::protoc_bin_path().map_err(|e| Box::new(e))?;
```

**Review by archerny — COMMENTED** [review]
This PR seems to introduce a potential vulnerability. I've checked the source code of this newly introduced dependency `rust-protoc-bin-vendored`. The protoc binary is self contained in the source repo, and version is locked to `libprotoc 3.19.4`. This version differs from the one which is used to generate `golang` and `java` code. How about we generate the rust code in advance and put it in the repo staticly.

File: rust/build.rs (hunk 1 above)
Comment (ShiKaiWi) [reply rust/build.rs:6]: I guess no need because it is possible that local protoc can't meet the requirement for the version.

**Comment by ShiKaiWi** [comment]
> This PR seems to introduce a potential vulnerability. I've checked the source code of this newly introduced dependency `rust-protoc-bin-vendored`. The protoc binary is self contained in the source repo, and version is locked to `libprotoc 3.19.4`. This version differs from the one which is used to generate `golang` and `java` code. How about we generate the rust code in advance and put it in the repo staticly.

I guess we prefer to generate code dynamically. Firstly, it is not a good idea to record the generated code in the git for that generated code will mess the every change set. Secondly, `protoc` version actually is not a serious problem if we keep them as the same major version.

**Pull request merged**